### PR TITLE
Add keyword completion kind

### PR DESCRIPTION
**Pull request by rgrinberg**
This will allow the LSP server to request keyword completion explicitly. This seems to be a highly desired feature: https://github.com/ocaml/ocaml-lsp/issues/173

### CODE REVIEW

**Comment by avsm** [comment]
This would really help the VSCode plugin, as the immediate keyword completion of `in` to `in_channel` is the first thing that gets in the way.

**Comment by jberdine** [comment]
Perhaps unrelated, but this makes me think that maybe this could help with avoiding the annoying behavior of `with` autocompleting to e.g. `with_return` in emacs.

File: src/frontend/query_commands.ml [hunk 1]
```diff
@@ -243,6 +243,13 @@ let reconstruct_identifier pipeline pos = function
         aux acc (succ i) in
     aux [] offset
 
+let keywords pipeline kinds =
+  match kinds with
+  | []
+  | _ :: _ when List.mem `Keywords ~set:kinds ->
```
Comment (trefis) [inline src/frontend/query_commands.ml:249]: That's a bit weird. Why not make it an `if`?

File: src/kernel/mreader.ml [hunk 2]
```diff
@@ -14,6 +14,7 @@ type result = {
   comments      : comment list;
   parsetree     : parsetree;
   no_labels_for_completion : bool;
+  keywords      : string list;
```
Comment (trefis) [inline src/kernel/mreader.ml:17]: I don't think this needs to be stored in there. That list is going to stay the same across all the parse / buffers / etc.

File: src/analysis/completion.mli [hunk 3]
```diff
@@ -47,6 +47,10 @@ val raw_info_printer : raw_info ->
   | `Concat of string * Extend_protocol.Reader.outcometree
   ]
 
+type keywords =
+  | Exclude
+  | Suggest of string list
```
Comment (trefis) [inline src/analysis/completion.mli:52]: Is there a good reason to pass a list of keywords as argument?
It seems we're always passing the same list (all the keywords), so perhaps the function should just be parametrized by `suggest_keywords:bool`?

File: src/ocaml/preprocess/411/lexer_raw.mll [hunk 4]
```diff
@@ -140,6 +140,10 @@ let keyword_table : keywords =
     "asr", INFIXOP4("asr");
 ]
 
+let ocaml_keywords =
+  let l = lazy (Hashtbl.fold (fun k _ acc -> k :: acc) keyword_table []) in
+  fun () -> Lazy.force l
```
Comment (trefis) [inline src/ocaml/preprocess/411/lexer_raw.mll:145]: Did you notice an impact on startup time if you remove the layer of lazy + closure? That is: if you just compute the list of keywords during module initialization.
There aren't that many of them, so hopefully it should be fine.

**Review by trefis — CHANGES_REQUESTED** [review]
This seems fine as a feature, but a couple of things need to happen before this gets merged:
- it needs rebasing (the blame is definitely on our side, sorry for not looking at this sooner). Hopefully this won't be too much work: most of the backends have disappeared!
- I don't think we need the keywords to bubble up through `Mreader_lexer`, `Mreader_parser`, etc. We can probably just access the lexer directly. And it's probably a bit cleaner if we get the list from `Completion` and not `Query_commands`
- I also left a couple of inline comments/questions

File: src/frontend/query_commands.ml (hunk 1 above)
Comment (rgrinberg) [reply src/frontend/query_commands.ml]: No longer present in the up to date patch

File: src/analysis/completion.mli (hunk 3 above)
Comment (rgrinberg) [reply src/analysis/completion.mli]: No longer present in the up to date patch

File: src/ocaml/preprocess/411/lexer_raw.mll (hunk 4 above)
Comment (rgrinberg) [reply src/ocaml/preprocess/411/lexer_raw.mll:145]: As you mentioned, it wouldn't work without custom keywords. Although it makes sense to remove custom keywords altogether? I don't think anyone uses camlp4 anymore, so I'm not sure what the functionality is useful for .

**Comment by rgrinberg** [comment]
@trefis I've updated the patch to the latest version actually used in lsp. I know you're working on your own patch, so this is just for your reference.

**Comment by voodoos** [comment]
I guess we can close this PR now that #1243 is merged.

**Pull request closed without merging**